src/platform_impl/web/cursor.rs

```diff
@@ -0,0 +1,278 @@
+use std::{
+    cell::RefCell,
+    mem,
+    ops::{Deref, DerefMut},
+    rc::{Rc, Weak},
+};
+
+use crate::cursor::{BadImage, CursorImage};
+use cursor_icon::CursorIcon;
+use wasm_bindgen::{closure::Closure, JsCast};
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{
+    Blob, Document, HtmlCanvasElement, ImageBitmap, ImageBitmapOptions,
+    ImageBitmapRenderingContext, ImageData, PremultiplyAlpha, Url, Window,
+};
+
+use super::backend::Style;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum WebCustomCursor {
+    Image(CursorImage),
+    Url {
+        url: String,
+        hotspot_x: u32,
+        hotspot_y: u32,
+    },
+}
+
+impl WebCustomCursor {
+    pub fn from_rgba(
+        rgba: Vec<u8>,
+        width: u32,
+        height: u32,
+        hotspot_x: u32,
+        hotspot_y: u32,
+    ) -> Result<Self, BadImage> {
+        Ok(Self::Image(CursorImage::from_rgba(
+            rgba, width, height, hotspot_x, hotspot_y,
+        )?))
+    }
+
+    pub(super) fn build(
+        &self,
+        window: &Window,
+        document: &Document,
+        style: &Style,
+        previous: SelectedCursor,
+    ) -> SelectedCursor {
+        match self {
+            WebCustomCursor::Image(image) => SelectedCursor::Image(CursorImageState::from_image(
+                window,
+                document.clone(),
+                style.clone(),
+                image,
+                previous,
+            )),
+            WebCustomCursor::Url {
+                url,
+                hotspot_x,
+                hotspot_y,
+            } => {
+                let value = format!("url({}) {} {}, auto", url, hotspot_x, hotspot_y);
+                style.set("cursor", &value);
+                SelectedCursor::Url(value)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum SelectedCursor {
+    Named(CursorIcon),
+    Url(String),
+    Image(Rc<RefCell<CursorImageState>>),
+}
+
+impl Default for SelectedCursor {
+    fn default() -> Self {
+        Self::Named(Default::default())
+    }
+}
+
+impl SelectedCursor {
+    pub fn set_style(&self, style: &Style) {
+        let value = match self {
+            SelectedCursor::Named(icon) => icon.name(),
+            SelectedCursor::Url(url) => url,
+            SelectedCursor::Image(image) => {
+                let image = image.borrow();
+                let value = match image.deref() {
```

Comment:
```suggestion
                let value = match &**image {
```

Reply:
I very much prefer using `deref()` instead of `&**`.
Why the change?

Reply:
Manual `deref()` is a code smell; there is a Clippy lint against it. It's much better to deref the data using the operator.

Reply:
I would say who cares because it's only enabled with pedantic and its reasoning isn't too convincing. https://rust-lang.github.io/rust-clippy/master/#/explicit_deref_methods

Reply:
> Manual `deref()` is a code smell; there is a Clippy lint against it. It's much better to deref the data using the operator.

Could you elaborate why it is a code smell? The Clippy lint is "pedantic" for a reason, imo it should just be "restriction".

Reply:
Fair enough; it's your backend, so go ahead.